src/app/utils/field-validation/validators/checkboxValidator.ts

```diff
@@ -0,0 +1,141 @@
+import { chain, left, right } from 'fp-ts/lib/Either'
+import { flow } from 'fp-ts/lib/function'
+
+import { ProcessedCheckboxResponse } from 'src/app/modules/submission/submission.types'
+import { ICheckboxField } from 'src/types/field'
+import { ResponseValidator } from 'src/types/field/utils/validation'
+
+type CheckboxValidator = ResponseValidator<ProcessedCheckboxResponse>
+type CheckboxValidatorConstructor = (
+  checkboxField: ICheckboxField,
+) => CheckboxValidator
+
+/**
+ * Returns a validator to check if answerArray is empty
+ */
+const checkboxAnswerValidator: CheckboxValidator = (response) => {
+  const { answerArray } = response
+
+  return answerArray.length === 0
+    ? left(`CheckboxValidator:\t Answer is empty array`)
+    : right(response)
+}
+
+/**
+ * Returns a validation function to check if number of
+ * selected checkbox options is less than the minimum number specified.
+ */
+const minOptionsValidator: CheckboxValidatorConstructor = (checkboxField) => (
+  response,
+) => {
+  const { validateByValue } = checkboxField
+  const { customMin } = checkboxField.ValidationOptions
+  const { answerArray } = response
+
+  return validateByValue && customMin
+    ? answerArray.length >= customMin
+      ? right(response)
+      : left(
+          `CheckboxValidator:\t answer has less options selected than minimum specified`,
+        )
+    : right(response)
+}
+
+/**
+ * Returns a validation function to check if number of
+ * selected checkbox options is more than the maximum number specified.
+ */
+const maxOptionsValidator: CheckboxValidatorConstructor = (checkboxField) => (
+  response,
+) => {
+  const { validateByValue } = checkboxField
+  const { customMax } = checkboxField.ValidationOptions
+  const { answerArray } = response
+
+  return validateByValue && customMax
+    ? answerArray.length <= customMax
+      ? right(response)
+      : left(
+          `CheckboxValidator:\t answer has more options selected than maximum specified`,
+        )
+    : right(response)
+}
+
+/**
+ * Returns a validation function to check if 'others' is selected
+ * and if so, whether the response for 'others' is valid.
+ */
+const othersOptionValidator: CheckboxValidatorConstructor = (checkboxField) => (
+  response,
+) => {
+  const { othersRadioButton } = checkboxField
+  const { answerArray } = response
+
+  if (othersRadioButton) {
```

Comment:
is this right? `othersRadioButton` just indicates whether the field has an "Others" option, not whether it was actually submitted. will this allow submission of a field without selecting the "Others" option?

Reply:
Thanks for pointing out; actually the state of 'others' option is not included in the response, only the value of others is. So it has always been possible to programatically submit a 'Others' response without selecting the others option.

On review, what is more serious is that the previous implementation would have blocked submission when 'Others' is enabled, but respondent selected something else and not others. This was not covered in the tests. I've rewritten the validation to address this, and added test case.

Reply:
>On review, what is more serious is that the previous implementation would have blocked submission when 'Others' is enabled, but respondent selected something else and not others.

yes this is what I meant

Reply:
sorry if I didn't explain it clearly